Libraries/LibWebView/EventLoop/EventLoopImplementationQt.cpp

```diff
@@ -13,30 +13,64 @@
 #include <LibCore/Notifier.h>
 #include <LibCore/System.h>
 #include <LibCore/ThreadEventQueue.h>
+#include <LibThreading/Mutex.h>
 #include <LibWebView/EventLoop/EventLoopImplementationQt.h>
 #include <LibWebView/EventLoop/EventLoopImplementationQtEventTarget.h>
 
 #include <QCoreApplication>
 #include <QEvent>
 #include <QEventLoop>
 #include <QSocketNotifier>
+#include <QThread>
 #include <QTimer>
 
 namespace WebView {
 
 struct ThreadData;
-static thread_local ThreadData* s_thread_data;
+static thread_local OwnPtr<ThreadData> s_this_thread_data;
+static HashMap<pthread_t, ThreadData*> s_thread_data;
+static thread_local pthread_t s_thread_id;
+static pthread_key_t s_thread_key;
+static pthread_rwlock_t s_thread_data_lock_impl;
+static pthread_rwlock_t* s_thread_data_lock = nullptr;
 
 struct ThreadData {
     static ThreadData& the()
     {
-        if (!s_thread_data) {
-            // FIXME: Don't leak this.
-            s_thread_data = new ThreadData;
+        if (!s_thread_data_lock) {
+            pthread_rwlock_init(&s_thread_data_lock_impl, nullptr);
+            s_thread_data_lock = &s_thread_data_lock_impl;
+            pthread_key_create(&s_thread_key, [](void*) {
+                s_this_thread_data.clear();
+            });
         }
-        return *s_thread_data;
+        if (s_thread_id == 0)
+            s_thread_id = pthread_self();
+        if (!s_this_thread_data) {
+            s_this_thread_data = make<ThreadData>();
+            pthread_rwlock_wrlock(&*s_thread_data_lock);
```

Comment:
(If we can't use RWLock) - de-referencing and then taking the address seems weird. Is this any different than just:
```c++
pthread_rwlock_wrlock(s_thread_data_lock);
```